Proposal: Implement additional SSR detection and control

KevinGrandon opened the pull request:
This implements two SSR detection features.

* **Additional non-SSR whitelisted extensions**
Adds the following extensions: (js|gif|jpg|png|pdf|json)

* **SSRDeciderToken for enhancing SSR logic**
Exports our current SSR logic as a new token, SSRDeciderToken, which can be enhanced through the DI system. This is a change which keeps our API surface fairly small, isn't difficult to support, and allows folks to control server rendering while we iterate on a future solution.

For more discussion and a breakdown of supported URIs see: https://github.com/fusionjs/fusion-core/issues/116#issuecomment-369430656

Fixes #45 
Fixes #116

ganemone commented on src/plugins/ssr.js:
```diff
@@ -1,9 +1,27 @@
+      // XHR/fetch requests do not have `text/html` in the Accept headers
+      if (!ctx.headers.accept) return false;
+      if (!ctx.headers.accept.includes('text/html')) return false;
+      return true;
```
yea we should add that check

lhorie commented:
Just yesterday, I was writing an analysis of various methods that had been proposed here: https://github.com/fusionjs/fusion-core/issues/116

This implements one of the recommendations. @slonoed's use case can potentially be better supported out-of-box via the other recommendation

With regards to this proposal, I think the injected service should be a function that can early return before the base algorithm. Replacing it altogether via enhance requires adding a lot of composition boilerplate for the most common case, and if said boilerplate is not implemented, you lose default semantics and might potentially break unrelated scenarios.

i.e.

```js
// do 
app.register(SSRDecider, (ctx) => !ctx.path.match(/paths|with|no|ssr/))

// instead of
const SSRDeciderEnhancer = ssrDecider =>
  createPlugin({
    provides: () => ctx => {
      return ssrDecider(ctx) && !ctx.path.match(/paths|with|no|ssr/);
    },
  });
app.enhance(SSRDeciderToken, SSRDeciderEnhancer);
```

KevinGrandon commented:
I did not see #116, but that is a very nice analysis. I think that this falls under the `Provide a isSSR DI hook for power users`, which can potentially allow people to completely override the existing SSR logic if necessary.  I like the enhance more than an early return fn because it's more powerful.

I think that if we couple this, along side extension checking int he default SSRDecider plugin, that will give us the best of both worlds. A basic extension check for most cases, and also enhancing for power users.

lhorie commented:
> I think that this falls under the Provide a isSSR DI hook for power users

Yeah, interesting that we came up with similar solutions independently :)

Something that just occurred to me: why do we need to return a plugin? Can we do:

```js
const SSRDeciderEnhancer = ssrDecider =>
  ctx => ssrDecider(ctx) && !ctx.path.match(/paths|with|no|ssr/);
```

Also, can we simplify the base check?

```
const SSRDeciderEnhancer = isSSRSuggested =>
  ctx => isSSRSuggested && !ctx.path.match(/paths|with|no|ssr/);
```

KevinGrandon commented:
> Something that just occurred to me: why do we need to return a plugin? Can we do:

Yes, I think I just like the idea that everything is a plugin, but we should be able to simplify it. We may also want to support the plugin case for deps though, but documenting the simple case is better for users. I will try to make those updates, along with test cases.

KevinGrandon commented:
New updates, simplified documented power-user enhancer. The new syntax is pretty clean:

```js
app.enhance(SSRDeciderToken, decide => ctx =>
  decide(ctx) && !ctx.path.match(/ignore-ssr-route/)
);
```

Also added additional extensions which should solve the majority of cases.

KevinGrandon commented:
!merge
